From: WuTong <wt@qkzy.net>
Date: Mon, 30 Mar 2026 17:30:00 +0800
Subject: [PATCH] ntfsplus: Fix iomap API for Linux 6.19

Linux 6.19 changed iomap_read_folio and iomap_readahead signatures:
- iomap_read_folio(struct folio *folio, const struct iomap_ops *ops)
+ iomap_read_folio(const struct iomap_ops *ops, struct iomap_read_folio_ctx *ctx)
- iomap_readahead(struct readahead_control *rac, const struct iomap_ops *ops)
+ iomap_readahead(const struct iomap_ops *ops, struct iomap_read_folio_ctx *ctx)

Signed-off-by: WuTong <wt@qkzy.net>
---
 fs/ntfsplus/aops.c | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/fs/ntfsplus/aops.c b/fs/ntfsplus/aops.c
index xxxxxxx..yyyyyyy 100644
--- a/fs/ntfsplus/aops.c
+++ b/fs/ntfsplus/aops.c
@@ -114,7 +114,14 @@ static int ntfs_read_folio(struct file *file, struct folio *folio)
 			return ntfs_read_compressed_block(folio);
 	}
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 19, 0)
+	struct iomap_read_folio_ctx ctx = {
+		.folio = folio,
+	};
+	iomap_read_folio(&ntfs_read_iomap_ops, &ctx);
+	return 0;
+#else
 	return iomap_read_folio(folio, &ntfs_read_iomap_ops);
+#endif
 }
 
 static int ntfs_write_mft_block(struct ntfs_inode *ni, struct folio *folio,
@@ -416,7 +423,16 @@ static void ntfs_readahead(struct readahead_control *rac)
 	    (ni->mft_no == FILE_MFT || ni->mft_no == FILE_MFTMirr))
 		return;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 19, 0)
+	struct iomap_read_folio_ctx ctx = {
+		.rac = rac,
+	};
+	iomap_readahead(&ntfs_read_iomap_ops, &ctx);
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(6, 16, 0)
 	iomap_readahead(rac, &ntfs_read_iomap_ops);
+#else
+	iomap_readahead(rac, &ntfs_read_iomap_ops);
+#endif
 }
 
 static int ntfs_mft_writepage(struct folio *folio, struct writeback_control *wbc)
-- 
2.48.1